rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,202 @@
+
+The `?VIEW_UPDATE_SEQ` row is used to keep track of what sequence in the database changes feed that the index has been indexed up to. It is not possible to update a map index in the same transaction that a document is updated, so the `VIEW_ID_RANGE` namespace is used to keep track of the emitted keys for a document for each map function in a design document. The ?VIEW_MAP_RANGE is the namespace storing the actual emitted keys and values from a map function for a document. For every emitted key/value from a map function, there are two rows in FoundationDB. The first row contains the emitted keys and the second row contains the emitted value from the map function, the `?ROW_KEYS` and `?ROW_VALUE` is added to the key so that emitted keys are always ordered before the value. When the emitted keys are encoded to binary to create the FDB key for a row, strings are converted to sort strings and all numbers are converted to doubles. It is not possible to convert either of those encoded values back to the original value. The `count` value is kept to allow for duplicates. Each emitted key/value pair for a document for a map function is given a `count`. 
+
+The `?VIEW_ROW_COUNT` is used to store the number of rows in the index and `?VIEW_KV_SIZE` keeps track of the size of this index. The size calculation is done using `erlang:external_size`.
+
+The process flow for a document to be indexed is as follows:
+
+1. FDB Transaction is started
+1. Read the document from the changes read (The number of documents to read at one type is configurable, the default is 100)
+1. The document is passed to the javascript query server and run through all the map functions defined in the design document
+1. The view's sequence number is updated to the sequence the document is in the changes feed.
+1. The emitted keys are stored in the `?VIEW_ID_RANGE`
+1. The emitted keys are encoded then added to the `?VIEW_MAP_RANGE` with the emitted keys and value stored
+1. The `?VIEW_ROW_COUNT` is incremented
+1. The `?VIEW_KV_SIZE` is increased
+1. If the document was deleted and was previously in the view, the previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE`. The Row count and size count are also decreased.
+1. If the document is being updated and was previously added to the index, then he previous keys for the document are read from `?VIEW_ID_RANGE` and then cleared from the `?VIEW_MAP_RANGE` and then the index is updated with the latest emitted keys and value.
+
+### Key ordering
+
+FoundationDB orders key by byte value which is not how CouchDB orders keys. To maintain CouchDB's view collation, a type value will need to be prepended to each key so that the correct sort order of null < boolean < numbers < strings < arrays < objects is maintained.
+
+In CouchDB 2.x, strings are compared via ICU. The way to do this with FoundationDB is that for every string an ICU sort string will be generated upfront and added to the key.
```

Comment:
By "added to the key" do you literally mean that the key will include the byte string and the ICU sort string byte string (this is what I read from "added") or that the key used within FoundationDB for the emitted key will only include the ICU sort string byte string?